src/extension/android/androidEmulatorManager.ts

```diff
@@ -0,0 +1,101 @@
+    constructor(adbHelper: AdbHelper) {
+        this.adbHelper = adbHelper;
+        this.childProcess = new ChildProcess();
+    }
+
+    public async startEmulator(target: string): Promise<IEmulator | null> {
+        if (target && (await this.adbHelper.getConnectedDevices()).length === 0) {
+            if (target === "simulator") {
+                const newEmulator = await this.selectEmulator();
+                if (newEmulator) {
+                    const emulatorId = await this.tryLaunchEmulatorByName(newEmulator);
+                    return {name: newEmulator, id: emulatorId};
+                }
+            }
+            else if (!target.includes("device")) {
+                const emulatorId = await this.tryLaunchEmulatorByName(target);
+                return {name: target, id: emulatorId};
+            }
+        }
+        return null;
+    }
+
+    public async tryLaunchEmulatorByName(emulatorName: string): Promise<string> {
+        return new Promise((resolve, reject) => {
+            const cp = this.childProcess.spawn(AndroidEmulatorManager.EMULATOR_COMMAND, [AndroidEmulatorManager.EMULATOR_AVD_START_COMMAND, emulatorName], {
+                detached: true,
+                stdio: 'ignore',
+              });
+            cp.spawnedProcess.unref();
+
+            const rejectTimeout = setTimeout(() => {
+                cleanup();
+                reject(`Could not start emulator within ${AndroidEmulatorManager.EMULATOR_START_TIMEOUT} seconds.`);
+            }, AndroidEmulatorManager.EMULATOR_START_TIMEOUT * 1000);
+
+            const bootCheckInterval = setInterval(async () => {
+                const connectedDevices = await this.adbHelper.getConnectedDevices();
+                if (connectedDevices.length > 0) {
+                    cleanup();
+                    resolve(connectedDevices[0].id);
+                }
+            }, 1000);
+
+            const cleanup = () => {
+                clearTimeout(rejectTimeout);
+                clearInterval(bootCheckInterval);
+            };
+        });
+    }
+
+    public  async selectEmulator(): Promise<string | undefined> {
+        const emulatorsList = await this.getEmulatorsList();
+        const quickPickOptions: QuickPickOptions = {
+            ignoreFocusOut: true,
+            canPickMany: false,
+            placeHolder: localize("SelectEmulator", "Select emulator for launch application"),
+        };
+        let result: string | undefined = emulatorsList[0];
+        if (emulatorsList.length > 1) {
+            result = await window.showQuickPick(emulatorsList, quickPickOptions);
+        }
+        return result?.toString();
+    }
+
+    private async getEmulatorsList(): Promise<string[]> {
+        const res = await this.childProcess.execToString(`${AndroidEmulatorManager.EMULATOR_COMMAND} ${AndroidEmulatorManager.EMULATOR_LIST_AVDS_COMMAND}`);
+        let emulatorsList: string[] = [];
+        if (res) {
+            emulatorsList = res.split("\r\n");
```

Comment:
This might not work on Unix-like systems like macOS. More information: https://en.wikipedia.org/wiki/Control_character
To avoid such control characters you can use `trim` function. See this for reference: https://github.com/Microsoft/vscode-react-native/blob/869b2b5c5f2d596ff9bf9ff769ee18ece11e5bfd/test/smoke/package/src/helpers/androidEmulatorHelper.ts#L69